backend/chunker/chunker.go

```diff
@@ -0,0 +1,1412 @@
+// that has had changes. If the implementation
+// uses polling, it should adhere to the given interval.
+func (f *Fs) ChangeNotify(ctx context.Context, notifyFunc func(string, fs.EntryType), pollIntervalChan <-chan time.Duration) {
+	do := f.Fs.Features().ChangeNotify
+	if do == nil {
+		return
+	}
+	wrappedNotifyFunc := func(path string, entryType fs.EntryType) {
+		//fs.Debugf(f, "ChangeNotify: path %q entryType %d", path, entryType)
+		if entryType == fs.EntryObject {
+			if match := f.nameRegex.FindStringSubmatchIndex(path); match != nil {
+				path = path[match[2]:match[3]]
+			}
+		}
+		notifyFunc(path, entryType)
+	}
+	do(ctx, wrappedNotifyFunc, pollIntervalChan)
+}
+
+// Object wraps one or more remote data chunks
+type Object struct {
+	remote string
+	main   fs.Object // meta object or single chunk
+	chunks []fs.Object
+	size   int64
+	f      *Fs
+}
+
+func (o *Object) addChunk(chunk fs.Object, chunkNum int) error {
+	if chunkNum < o.f.opt.StartFrom {
+		return fmt.Errorf("invalid chunk number %d", chunkNum)
+	}
+	idx := chunkNum - o.f.opt.StartFrom
+	if idx == len(o.chunks) {
+		o.chunks = append(o.chunks, chunk)
+		return nil
+	}
+	if idx > len(o.chunks) {
+		newChunks := make([]fs.Object, (idx + 1), (idx+1)*2)
+		copy(newChunks, o.chunks)
+		o.chunks = newChunks
+	}
+	o.chunks[idx] = chunk
+	return nil
+}
+
+func (o *Object) validate() error {
+	if o.main == nil && o.chunks == nil {
+		panic("invalid chunked object")
+	}
+	if o.chunks == nil {
+		// non chunked object
+		return nil
+	}
+	// this object is chunked
+	if o.main != nil && o.main.Size() > maxMetaDataSize {
+		// metadata of a chunked object must be a tiny piece of json
+		o.size = -1
+		return fmt.Errorf("%q metadata is too large", o.remote)
+	}
+	var totalSize int64
+	for _, chunk := range o.chunks {
+		if chunk == nil {
+			o.size = -1
+			return fmt.Errorf("%q misses chunks", o)
+		}
+		totalSize += chunk.Size()
+	}
+	o.size = totalSize
+	return nil
+}
+
+func (f *Fs) newObject(remote string, main fs.Object, chunks []fs.Object) *Object {
+	var size int64 = -1
+	if main != nil {
+		size = main.Size()
+		if remote == "" {
+			remote = main.Remote()
+		}
+	}
+	return &Object{
+		remote: remote,
+		main:   main,
+		size:   size,
+		f:      f,
+		chunks: chunks,
+	}
+}
+
+// Fs returns read only access to the Fs that this object is part of
+func (o *Object) Fs() fs.Info {
+	return o.f
+}
+
+// Return a string version
+func (o *Object) String() string {
+	if o == nil {
+		return "<nil>"
+	}
+	return o.remote
+}
+
+// Remote returns the remote path
+func (o *Object) Remote() string {
+	return o.remote
+}
+
+// Size returns the size of the file
+func (o *Object) Size() int64 {
+	if o.chunks != nil {
+		return o.size
+	}
+	return o.main.Size()
+}
+
+// Storable returns whether object is storable
+func (o *Object) Storable() bool {
+	if o.main != nil {
+		return o.main.Storable()
+	}
+	if o.chunks == nil {
+		panic("invalid chunked object")
```

Comment:
Is this likely?

Maybe factor this into a method on object `firstChunk` since it seems to be used several times?

Reply:
Generally unlikely bug possible. Factoring this out seems a good idea.